src/Analyser/NodeScopeResolver.php

```diff
@@ -5625,10 +5625,27 @@ static function (): void {
 				$assignedExprType = $scope->getType($assignedExpr);
 				$nodeCallback(new PropertyAssignNode($var, $assignedExpr, $isAssignOp), $scope);
 				if ($propertyReflection->canChangeTypeAfterAssignment()) {
-					if ($propertyReflection->hasNativeType() && $scope->isDeclareStrictTypes()) {
+					if ($propertyReflection->hasNativeType()) {
+						$assignedNativeType = $scope->getNativeType($assignedExpr);
 						$propertyNativeType = $propertyReflection->getNativeType();
 
-						$scope = $scope->assignExpression($var, TypeCombinator::intersect($assignedExprType->toCoercedArgumentType(true), $propertyNativeType), TypeCombinator::intersect($scope->getNativeType($assignedExpr)->toCoercedArgumentType(true), $propertyNativeType));
+						$assignedTypeIsCompatible = false;
+						foreach (TypeUtils::flattenTypes($propertyNativeType) as $type) {
+							if ($type->isSuperTypeOf($assignedNativeType)->yes()) {
+								$assignedTypeIsCompatible = true;
+								break;
+							}
+						}
+
+						if (!$assignedTypeIsCompatible && $scope->isDeclareStrictTypes()) {
+							$scope = $scope->assignExpression(
+								$var,
+								TypeCombinator::intersect($assignedExprType->toCoercedArgumentType(true), $propertyNativeType),
+								TypeCombinator::intersect($assignedNativeType->toCoercedArgumentType(true), $propertyNativeType),
+							);
+						} else {
+							$scope = $scope->assignExpression($var, $assignedExprType, $assignedNativeType);
```

Comment:
This is gonna get executed if assigned type is compatible or if we're not in declare strict types. That's not what we want. We only want to execute this if assigned type is compatible.

Reply:
so what to execute if not compatible and not strict types

Reply:
I think it works now as you imagined.

Reply:
If not compatible and not strict types, we currently can't do anything. Once we properly implement toCoercedArgumentType in the whole typesystem, then we can use it same way we use it now with strict_types=1.